### PR TITLE
Feature/orders

### PR DESCRIPTION
**Pull request by stepan-zubkov**
# Orders!

Screenshot_20221010_213918

- Add models for orders
- Add integration with PSP Fondy
- Add user's orders display in profile
- Add tests for orders

### CODE REVIEW

File: butterfly/orders/payment.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    """
+    data = {
+        "order_id": str(order.unique_id),
+        "order_desc": generate_order_desc(order),
+        "currency": "RUB",
+        "amount": str(int(order.get_amount() * 100)),
+        "response_url": request.build_absolute_uri(reverse("orders:approve_payment")),
+    }
+    signature = generate_signature(params=data)
+    data = {**data, "signature": signature}
```
Comment (kirillzhosul) [inline butterfly/orders/payment.py:27]: data.update({"signature": signature})

File: butterfly/orders/payment.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    Returns:
+        str: Description
+    """
+    count = order.items.count()
+    if count == 1:
+        return order.items.all()[0].product.name
+    elif count == 2:
+        return ", ".join([i.product.name for i in order.items.all()])
+    else:
```
Comment (kirillzhosul) [inline butterfly/orders/payment.py:80]: Remove else

File: butterfly/orders/views.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        cart.delete()
+
+        api = Api(
+            merchant_id=settings.FONDY_MERCHANT_ID,
+            secret_key=settings.FONDY_SECRET_KEY,
+        )
+        checkout = Checkout(api=api)
+
+        data = generate_order_data(order, request)
+        url = checkout.url(data).get("checkout_url")
```
Comment (kirillzhosul) [inline butterfly/orders/views.py:44]: Maybe there is should be some business-logic service where you can encapsulate that?

File: butterfly/orders/views.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+def approve_payment(request):
+    if request.method == "POST":
+        payment_data = request.POST
+        order = get_object_or_404(Order, unique_id=payment_data.get("order_id"))
+        if check_signature(request):
+            order.status = payment_data["order_status"]
+            order.save()
+            return redirect(reverse("users:redirect"))
+
+        raise Http404()
```
Comment (kirillzhosul) [inline butterfly/orders/views.py:60]: Remove first raise

File: config/settings/base.py [hunk 5]
```diff
@@ -281,3 +280,5 @@
 
 # Your stuff...
 # ------------------------------------------------------------------------------
+FONDY_MERCHANT_ID = env.int("FONDY_MERCHANT_ID", default=1397120)
+FONDY_SECRET_KEY = env.str("FONDY_SECRET_KEY", default="test")
```
Comment (kirillzhosul) [inline config/settings/base.py:284]: Is this safe env vars defaults?

File: requirements/base.txt [hunk 6]
```diff
@@ -15,3 +15,4 @@ django-allauth==0.51.0  # https://github.com/pennersr/django-allauth
 django-crispy-forms==1.14.0  # https://github.com/django-crispy-forms/django-crispy-forms
 crispy-bootstrap5==0.6  # https://github.com/django-crispy-forms/crispy-bootstrap5
 django-redis==5.2.0  # https://github.com/jazzband/django-redis
+cloudipsp
```
Comment (kirillzhosul) [inline requirements/base.txt:18]: No version, source link

**Review by kirillzhosul — APPROVED** [review]

File: config/settings/base.py (hunk 5 above)
Comment (stepan-zubkov) [reply config/settings/base.py:284]: Yes, these defaults are safe. It's only testing data, published on fondy api docs

Comment (kirillzhosul) [reply config/settings/base.py:284]: Just saw that there is comment above that looks... pretty weird...

File: requirements/base.txt (hunk 6 above)
Comment (stepan-zubkov) [reply requirements/base.txt:18]: This package is partly broken. See cloudipsp/python-sdk#5

File: butterfly/orders/payment.py (hunk 1 above)
Comment (stepan-zubkov) [reply butterfly/orders/payment.py:27]: Should be changed

File: butterfly/orders/payment.py (hunk 2 above)
Comment (stepan-zubkov) [reply butterfly/orders/payment.py:80]: Should be changed

File: butterfly/orders/views.py (hunk 3 above)
Comment (stepan-zubkov) [reply butterfly/orders/views.py:44]: Как тут энкапсулировать и вообще надо ли оно тут

File: butterfly/orders/views.py (hunk 4 above)
Comment (stepan-zubkov) [reply butterfly/orders/views.py:60]: No!

Comment (stepan-zubkov) [reply butterfly/orders/views.py:60]: Ok, i will remove...

**Review by PavelGrebenshikov — APPROVED** [review]

**Pull request merged**